homeassistant/components/generic/diagnostics.py

```diff
@@ -0,0 +1,49 @@
+"""Diagnostics support for generic (IP camera)."""
+from __future__ import annotations
+
+from typing import Any
+
+import yarl
+
+from homeassistant.components.diagnostics import async_redact_data
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+
+from .const import CONF_STILL_IMAGE_URL, CONF_STREAM_SOURCE
+
+TO_REDACT = {
+    CONF_PASSWORD,
+    CONF_USERNAME,
+}
+
+
+# A very similar redact function is in components.sql.  Possible to be made common.
+def redact_url(data: str) -> str:
+    """Redact credentials from string url."""
+    url_in = yarl.URL(data)
+    if url_in.user:
+        url = url_in.with_user("****")
+    if url_in.password:
+        url = url.with_password("****")
+    if url_in.path:
+        url = url.with_path("****")
+    if url_in.query_string:
+        url = url.with_query("****=****")
+    return str(url)
```

Comment:
Any reason to use `url_in` and `url`?  Can it just be `url`?

Reply:
I found that `url.with_path()` erases the query_string for some reason.  So without using `url_in`, the function behaves like the query string is always blank.

I copied it to the others to look consistent.  But it is only needed on L29